### PR TITLE
lora : raise error if lm_head is ignored

### LINKED ISSUE #9065: Bug: Gemma2 adapter weights `lm_head` skipped on gguf conversion
### What happened?

The `lm_head` layer for a [Gemma2](https://huggingface.co/google/gemma-2-2b) LoRA adapter is not converted by `convert_lora_to_gguf.py`, and therefore not applied at inference (ruining performance of the adapter).

<br>

### How to reproduce:

<details>
<summary>Expand</summary>

<br>

1. LoRA fine-tune Gemma2 with `pytorch`/`peft` including `lm_head` in the `target_modules` param:
    ```python
    config = LoraConfig(target_modules=["lm_head"], ...)
    ``` 
2. Save the adapter.
3. Convert the adapter debugging  
    ```bash
    python convert_lora_to_gguf.py <adapter folder> --base <base model folder> --outtype f32
    ```
    then the `lm_head` layer is skipped by [this line in `convert_hf_to_gguf.py`](https://github.com/ggerganov/llama.cpp/blob/4b9afbbe9037f8a2d659097c0c7d9fce32c6494c/convert_hf_to_gguf.py#L2648) (and no error is raised):
    ```python
    if name == "lm_head.weight":
       logger.debug(f"Skipping get tensor {name!r} in safetensors so that convert can end normally.")
       return []
    ```
4. Run `llama-cli` to check that indeed no lora layer is applied in the [respective line in llama.cpp](https://github.com/ggerganov/llama.cpp/blob/8b3befc0e2ed8fb18b903735831496b8b0c80949/src/llama.cpp#L12021):
    ```bash
    ./llama-cli -m base/model/path/Base-F32.gguf \
    --lora lora/model/path/Lora-F32-LoRA.gguf \
    -p "Hello Gemma2" -n 50
    ```

</details>

<br>

### Expected behaviour

I think this is a bug because a user might have trained an adapter that is applied to the the `lm_head` layer, so skipping it on conversion will destroy the adapter's performance. I think the code should either:
- **raise** an error saying `Cannot convert Gemma2 adapter with lm_head layer`

or

- **handle the lm_head layer** (although it might be tricky for merging adapters as the `lm_head` layer shares the weights with the `embed` layer in Gemma2, probably leading to having to create a new tensor for the `lm_head` to merge the adapter to).
<br>

### Comments

- I think the script `convert_lora_to_gguf.py` was introduced in PR #8332, so maybe the @ngxson knows if skipping the `lm_head` is the desired outcome of if it is actually a bug. Otherwise I'm happy to try figure out why this happens.
- This is not the case for, say, Phi3, which converts the `lm_head` lora layer correctly.
- I can provide more code/models to reproduce the bug easily if that helps.

<br>

### Name and Version

version: 3524 (bc0f887e)
built with Apple clang version 15.0.0 (clang-1500.3.9.4) for arm64-apple-darwin23.4.0

### What operating system are you seeing the problem on?

MacOS, but it should be a platform-independent problem.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ngxson**
Resolve #9065 

We will now raise an error if `lm_head` is ignored in base model.

TODO: @ltoniazzi can you test this? Thanks.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

File: convert_lora_to_gguf.py [hunk 1]
```diff
@@ -363,7 +363,11 @@ def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
                     yield (name, cast(torch.Tensor, LoraTorchTensor(tensor.A, tensor.B)))
 
             def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
-                dest = super().modify_tensors(data_torch, name, bid)
+                dest = list(super().modify_tensors(data_torch, name, bid))
```
Comment (ltoniazzi) [inline convert_lora_to_gguf.py:366]: Is it needed to recast to a list `super().modify_tensors(data_torch, name, bid)`?

File: convert_lora_to_gguf.py [hunk 2]
```diff
@@ -363,7 +363,11 @@ def get_tensors(self) -> Iterator[tuple[str, Tensor]]:
...
+                # for now, we cannot convert archs that use the same tensor for tok_embd and output
```
Comment (ltoniazzi) [inline convert_lora_to_gguf.py:367]:
```suggestion
                # for now, we cannot convert adapters with lm_head for archs that use the same tensor for tok_embd and output
```

**Review by ltoniazzi — COMMENTED** [review]
Looks good! 

Tested it on Gemma2 with (raise) and without (succeed) `lm_head` in adapter.

File: convert_lora_to_gguf.py (hunk 1 above)
Comment (ngxson) [reply convert_lora_to_gguf.py:366]: Without this, pyright will not pass: https://github.com/ggerganov/llama.cpp/actions/runs/10472198558/job/29001100132

Comment (compilade) [reply convert_lora_to_gguf.py:366]: Yes, because it can also be a `Generator` (when `modify_tensors` uses `yield`)

**Review by compilade — APPROVED** [review]

File: convert_lora_to_gguf.py (hunk 2 above)
Comment (ngxson) [reply convert_lora_to_gguf.py:367]: I’ll rephrase this to void mixing naming scheme
